pydantic_ai_slim/pydantic_ai/models/__init__.py

```diff
@@ -546,6 +546,86 @@ def prompted_output_instructions(self) -> str | None:
     __repr__ = _utils.dataclasses_no_defaults_repr
 
 
+@dataclass
+class ResolvedToolChoice:
+    """Provider-agnostic resolved tool choice.
+
+    This is the result of validating and resolving the user's `tool_choice` setting.
+    Providers should map this to their API-specific format.
+    """
+
+    mode: Literal['none', 'auto', 'required', 'specific']
+    """The resolved tool choice mode."""
+
+    tool_names: list[str] = field(default_factory=list)
+    """For 'specific' mode, the list of tool names to force. Empty for other modes."""
+
+    output_tools_fallback: bool = False
+    """True if we need to fall back to output tools only (when 'none' was requested but output tools exist)."""
+
+
+_TOOL_CHOICE_NONE_WITH_OUTPUT_TOOLS_WARNING = (
+    "tool_choice='none' is set but output tools are required for structured output. "
+    'The output tools will remain available. Consider using `NativeOutput` or `PromptedOutput` '
+    "if you need tool_choice='none' with structured output."
+)
+
+
+# NOTE: for PR discussion: should this be a private method? a Model.method? Perhaps a ModelRequestParameters.method?
+def resolve_tool_choice(
```

Comment:
Let's make it a private method for now, but I could see it eventually making sense to do this directly in `prepare_request`